File: Haskell/src/UI/hud.hs
module Haskell.UI.HUD (mainScreen, saveJogoScreen, mercadoScreen) where

-- Tela Principal da HUD 

mainScreen :: IO()
mainScreen = do
    let jogador = "Neymar"
    putStrLn ("Jogador: " ++ jogador)
    putStrLn ""
    putStrLn ""
    imprimiTabelas 0
    putStrLn ""
    putStrLn ""
    let inimigosDerrotados = 4
    let inimigosTotais = 10
    putStrLn ("Inimigos: " ++ show inimigosDerrotados ++ "/" ++ show inimigosTotais)


    let espacosAtingidos = 3
    let espacosTotais = 12
    putStrLn ("Espaços especiais: " ++ show espacosAtingidos ++ "/" ++ show espacosTotais)


    let minasOuroAtingidas = 2
    let minasOuroTotais = 5
    putStrLn ("Minas de ouro: " ++ show minasOuroAtingidas ++ "/" ++ show minasOuroTotais)
    
    putStrLn ""

    mostraInventario
    
    putStrLn ""

    putStrLn "Atalhos: "
    putStrLn "'1' -> Usar bombas pequenas"
    putStrLn "'2' -> Usar bombas médias"
    putStrLn "'3' -> Usar bombas grandes"
    putStrLn "{COLUNAS}{LINHA} -> Coordenada que deseja atacar; Exemplo: C3"
    putStrLn "'4' -> Usar o drone visualizador de áreas"
    putStrLn "'m' -> Acesso ao mercado"
    putStrLn "'s' -> Salvar o jogo no estado atual"
    putStrLn "'quit' -> Sair do jogo sem salvar"
    
    putStrLn ""

    putStrLn "> Digite a opção ou coordenadas:"

-- Tela Salvar Jogo

saveJogoScreen :: IO ()
saveJogoScreen = do
    putStrLn "Escolha um slot para salvar:"
    
    putStrLn ""

    putStrLn "[1] Slot 1 - Vazio"
    putStrLn "{nome jog.} - Jogo Salvo em {data}"
    putStrLn "{nome jog.} - Jogo Salvo em {data}"
    
    putStrLn ""

    putStr "> Digite o slot ou 'v' para voltar: "

-- Tela Mercado 

mercadoScreen :: IO ()
mercadoScreen = do

    putStrLn "  __  __                             _        "
    putStrLn " |  \x5C/  |  ___  _ __  ___  __ _   __| |  ___  "
    putStrLn " | |\x5C/| | / _ \x5C| '__|/ __|/ _` | / _` | / _ \x5C "
    putStrLn " | |  | ||  __/| |  | (__| (_| || (_| || (_) |"
    putStrLn " |_|  |_| \x5C___||_|   \x5C___|\x5C__,_| \x5C__,_| \x5C___/ "

    putStrLn ""

    mostraInventario
    
    putStrLn ""

    putStrLn "Atalhos:"
    
    putStrLn ""

    putStrLn "'1' -> Comprar bombas pequenas             ({preço})"
    putStrLn "'2' -> Comprar bombas médias               ({preço})"
    putStrLn "'3' -> Comprar bombas grandes              ({preço})"
    putStrLn "'4' -> Comprar drone visualizador de áreas ({preço})"

    putStrLn ""

    putStr "> Digite o item ou 'v' para voltar:"

-- Funções utilitárias

mostraInventario :: IO ()
mostraInventario = do
    putStrLn "Inventário:"
    
    let bombasPequenas = 3
    let bombasMedias = 1
    let bombasGrandes = 0
    putStrLn ("Bombas pequenas: " ++ show bombasPequenas ++ " | Bombas médias: " ++ show bombasMedias ++ " | Bombas grandes: " ++ show bombasGrandes)

    let drones = 2
    putStrLn ("Drone visualizador de áreas: " ++ show drones)

    putStrLn ""
    
    let moedas = 25
    putStrLn ("Moedas: " ++ show moedas)


imprimiTabelas :: Int -> IO() 
imprimiTabelas x = if x >= 11
    then do
        let elemTabela = unwords (geraTabela !! x)
        putStrLn (elemTabela ++ "                             " ++ elemTabela)
    else do
        if x < 10 
            then do
                let elemTabela = " " ++ unwords (geraTabela !! x)
                putStrLn (elemTabela ++ "                             " ++ elemTabela)
                imprimiTabelas (x + 1)
            else do
                let elemTabela = unwords (geraTabela !! x)
                putStrLn (elemTabela ++ "                             " ++ elemTabela)
                imprimiTabelas (x + 1)


geraTabela :: [[String]]
geraTabela = geraTabelaAux [["  ", "A ", "B ", "C ", "D ", "E ", "F ", "G ", "H ", "I ", "J ", "K ", "L "]]
-- geraTabela = geraTabelaAux ([[" "] ++ geraPrimeiraLinha ['A'..'L']])


geraTabelaAux :: [[String]] -> [[String]]
geraTabelaAux tabela = 
    if length tabela >= 13
        then [head tabela] ++ adicionaNumeroCadaLinha (tail tabela)
        else geraTabelaAux (tabela ++ [geraLinhasComuns [1..12]])

    

-- geraPrimeiraLinha :: [Char] -> [String]
-- geraPrimeiraLinha listaChar = [show x ++ " " | x <- listaChar]


geraLinhasComuns :: [Int] -> [String]
geraLinhasComuns [] = []
geraLinhasComuns (a : as) = ["X "] ++ (geraLinhasComuns as)


adicionaNumeroCadaLinha :: [[String]] -> [[String]]
adicionaNumeroCadaLinha listaSemNumeros = [ [show i ++ " "] ++ x | (x, i) <- zip listaSemNumeros [1..] ]

File: Prolog/src/UI/HUD.pl
:- module(hud, [mainScreen/1, saveJogoScreen/0, mercadoScreen/1]).

:- use_module('./UtilsUI').

:- use_module('../Modelos/Jogo').
:- use_module('../Modelos/Tabuleiro').
:- use_module('../Modelos/Jogador').
:- use_module('../Modelos/Bot').

% Regra principal da HUD

mainScreen(Jogo) :-
    getJogador(Jogo, Jogador),
    getNome(Jogador, Nome),
    getTabelaJogador(Jogador, TabelaJogador),
    getBot(Jogo, Bot),
    getTabelaBot(Bot, TabelaBot),
    format('Jogador: ~w', [Nome]),
    write('                           '),
    write('Oponente: República de Haskelland'), nl, nl,
    mainTabela(TabelaJogador, TabelaBot), nl, nl,
    inimigosDestruidos(TabelaJogador, TabelaBot), nl,
    espacosAmigosAtingidos(TabelaJogador, TabelaBot), nl, nl,
    inventario(Jogador), nl, nl,
    moedas(Jogador), nl, nl,
    writeln('Atalhos:'),
    writeln('\'1\' -> Usar bomba pequena'),
    writeln('\'2\' -> Usar bomba média'),
    writeln('\'3\' -> Usar bomba grande'),
    writeln('\'4\' -> Usar o drone visualizador de áreas'),
    writeln('{COLUNA}{LINHA} -> Coordenada que deseja atacar; Exemplo: C3'),
    nl,
    writeln('\'m\' -> Acesso ao mercado'),
    writeln('\'s\' -> Salvar o jogo no estado atual'),
    writeln('\'q\' -> Sair do jogo sem salvar'),
    nl,
    write('> Digite a opção: ').

% Regras auxiliares para a tabela

inteiroParaString(N, S3) :-
    atom_number(A, N), 
    atom_string(A, S),
    string_concat(S, ' ', S2),
    (N < 10 -> string_concat(' ', S2, S3) ; S3 = S2).

adicionaNumeros(X, Y, _, []) :- Y is X + 1, !.
adicionaNumeros(NumLinhas, I, [H | T], [R | C]) :-
    I =< NumLinhas,
    I2 is I + 1,
    inteiroParaString(I, S),
    append([S], H, R),
    adicionaNumeros(NumLinhas, I2, T, C).

concatenaCabecalho(T, NewT) :-
    append([["   ", "A ", "B ", "C ", "D ", "E ", "F ", "G ", "H ", "I ", "J ", "K ", "L "]], T, NewT).

% Nova regra para imprimir duas tabelas lado a lado

imprimiTabelasLadoALado([], []) :- !.
imprimiTabelasLadoALado([H1 | T1], [H2 | T2]) :-
    atomic_list_concat(H1, '', Str1),
    atomic_list_concat(H2, '', Str2),
    format('~w         ~w~n', [Str1, Str2]),
    imprimiTabelasLadoALado(T1, T2).

% Nova regra principal

mainTabela(TabelaJogador, TabelaBot) :-
    geraTabuleiroString(TabelaJogador, TabelaStrJogador),
    geraTabuleiroString(TabelaBot, TabelaStrBot),

    adicionaNumeros(12, 1, TabelaStrJogador, NumJogador),
    adicionaNumeros(12, 1, TabelaStrBot, NumBot),

    concatenaCabecalho(NumJogador, TabelaFinalJogador),
    concatenaCabecalho(NumBot, TabelaFinalBot),

    imprimiTabelasLadoALado(TabelaFinalJogador, TabelaFinalBot).


% Tela salvar jogo

saveJogoScreen :-
    writeln('Escolha um slot para salvar:'), nl,
    saveStates(Estados),
    write(Estados). 

% Inventário

inventario(Jogador) :-
    getBombasPequenas(Jogador, BP),
    getBombasMedias(Jogador, BM),
    getBombasGrandes(Jogador, BG),
    getDroneVisualizador(Jogador, DV),
    writeln('Inventário:'),
    format('Bombas pequenas: ~w | Bombas médias: ~w | Bombas grandes: ~w', [BP, BM, BG]), nl,
    format('Drone visualizador de áreas: ~w', [DV]).

moedas(Jogador) :-
    getMoedas(Jogador, Moedas),
    format('Moedas: ~w', [Moedas]).

% Mercado

mercadoScreen(Jogo) :-
    logoMercado,
    getJogador(Jogo, Jogador),
    inventario(Jogador),
    nl, write('Atalhos:'), nl,
    format('\'1\' -> Comprar bombas médias               ($~d)~n', [250]),
    format('\'2\' -> Comprar bombas grandes              ($~d)~n', [400]),
    format('\'3\' -> Comprar drone visualizador de áreas ($~d)~n', [350]),
    nl, write('> Digite o item ou \'v\' para voltar: ').

% Regras auxiliares da HUD

inimigosDestruidos(TabelaJogador, TabelaBot) :-
    contabilizarInimigos(TabelaJogador, InimigosJogador),
    contabilizarInimigos(TabelaBot, InimigosBot),
    format('Inimigos: ~w/6', [InimigosJogador]),
    write('                        '),
    format('Inimigos do Oponente: ~w/6', [InimigosBot]).

espacosAmigosAtingidos(TabelaJogador, TabelaBot) :-
    contabilizarAmigos(TabelaJogador, AmigosJogador),
    contabilizarAmigos(TabelaBot, AmigosBot),
    format('Espaços Amigos: ~w/3', [AmigosJogador]),
    write('                 '),
    format('Espaços Amigos do Oponente: ~w/3', [AmigosBot]).

logoMercado :-
    writeln('  __  __                             _        '),
    writeln(' |  \\/  |  ___  _ __  ___  __ _   __| |  ___  '), 
    writeln(' | |\\/| | / _ \\| \'__|/ __|/ _` | / _` | / _ \\ '),
    writeln(' | |  | ||  __/| |  | (__| (_| || (_| || (_) |'),
    writeln(' |_|  |_| \\___||_|   \\___|\\__,_| \\__,_| \\___/ '),
    writeln('').
